### ISSUE TITLE
mwoffliner broken with Parsoid 0.9.0

### ISSUE THREAD

**Issue by kelson42** [openzim/mwoffliner#350]
I have updated the package.json to use last parsoid version 0.9.0.

Unfortunately, it fails at the init stage with the following errror
```
$ ./bin/mwoffliner.script.js --mwUrl="https://www.dandwiki.com/" --adminEmail=admin@kiwix.com --localParsoid --verbose
Starting Parsoid
info loading ...
Error starting Parsoid: TypeError: Cannot read property 'constructor' of undefined
```

**Comment by kelson42** [comment]
@subbuss Any clue? With 0.8.0, it works fine.

**Comment by kelson42** [comment]
@ISNIT0 Would you be able to have a look to this?

**Comment by cscott** [comment]
I believe the latest Parsoid requires node 6 (or above).  Stable debian distros might be stuck on node 4 or so?

I (always) recommend the use of [nave](https://github.com/isaacs/nave) which makes it easy to temporarily switch to whatever node version you like/need.

**Comment by kelson42** [comment]
@cscott This is not the problem, it was at least version 8.

**Comment by kelson42** [comment]
@ISNIT0 It would be really great if you could have a look to this regression. This stops us the use the newest version of mwoffliner locally (so we keep an old parsoid for all non-wikimedia wikis).

**Comment by ISNIT0** [comment]
I'll take a look this evening

**Comment by ISNIT0** [comment]
@cscott 
Looks like there is an issue (or maybe it's deliberate) in Parsoid.
When no `logging` or `metrics` config is passed to `parsoid.apiServiceWorker`, something internal breaks because of a lack of `options.name`.

**Comment by kelson42** [comment]
@ISNIT0 Are you sure your patch works, for me the following command fails:
```
mwoffliner --withZimFullTextIndex --adminEmail=contact@kiwix.org --deflateTmpHtml --verbose --skipHtmlCache --skipCacheCleaning --format= --format=novid --format=nopic --mwUrl=https://wiki.installgentoo.com --localParsoid --mwApiPath=api.php --mwWikiPath=index.php
```

The errors look like:
```
Absolutely unable to retrieve async. URL: Unable to download content [3] http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/Bash_personalization?oldid=30660 (response code: 406).
SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at /usr/local/lib/node_modules/mwoffliner/lib/mwoffliner.lib.js:1492:29
    at async.retry (/usr/local/lib/node_modules/mwoffliner/lib/Downloader.js:176:7)
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:676:51
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:726:13
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:52:16
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:264:21
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:44:16
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:723:17
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:167:37
    at Timeout._onTimeout (/usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:652:25)
    at ontimeout (timers.js:429:13)
    at tryOnTimeout (timers.js:289:5)
    at listOnTimeout (timers.js:252:5)
    at Timer.processTimers (timers.js:212:10)
```

**Comment by arlolra** [comment]
> Looks like there is an issue (or maybe it's deliberate) in Parsoid.  When no logging or metrics config is passed to parsoid.apiServiceWorker, something internal breaks because of a lack of options.name.

That API is expected to be used in conjunction with service-runner,
https://github.com/wikimedia/service-runner

Your best bet, based on,
https://github.com/wikimedia/parsoid/blob/master/lib/index.js#L34-L56

would be to set `parent` so that default behaviour is overridden,

```
config: {
  localsettings: '../../localsettings.js',
  parent: undefined,
}
```

**Comment by ISNIT0** [comment]
It's working for me, I get a lot of files downloaded.
I think @arlolra's suggestion is best, so I'll implement that.

**Comment by cscott** [comment]
I believe @arlolra is looking at this, he may have made some changes recently that may be involved.

**Comment by kelson42** [comment]
@ISNIT0 To me unfortunately it still fails (same command as given before)... Error seem to be similar... but not sur

```
[fatal/request][wiki$0/Gentoo] Not acceptable.
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/1.6.1"
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/2.0.0"

Downloading http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/IRC?oldid=33427...
Downloading http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/Mechanical_keyboards?oldid=34530...
Unable to download content [2] http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/Web_browsers?oldid=34893 (response code: 406).
Unable to download content [1] http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/Gentoo?oldid=33819 (response code: 406).
[fatal/request][wiki$0/IRC] Not acceptable.
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/1.6.1"
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/2.0.0"

[fatal/request][wiki$0/Mechanical_keyboards] Not acceptable.
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/1.6.1"
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/2.0.0"

Unable to download content [1] http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/IRC?oldid=33427 (response code: 406).
Unable to download content [1] http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/Mechanical_keyboards?oldid=34530 (response code: 406).
[fatal/request][wiki$0/Web_browsers] Not acceptable.
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/1.6.1"
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/2.0.0"

[fatal/request][wiki$0/Gentoo] Not acceptable.
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/1.6.1"
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/2.0.0"

Unable to download content [3] http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/Web_browsers?oldid=34893 (response code: 406).
Unable to download content [2] http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/Gentoo?oldid=33819 (response code: 406).
Absolutely unable to retrieve async. URL: Unable to download content [3] http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/Web_browsers?oldid=34893 (response code: 406).
SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at /usr/local/lib/node_modules/mwoffliner/lib/mwoffliner.lib.js:1487:29
    at async.retry (/usr/local/lib/node_modules/mwoffliner/lib/Downloader.js:176:7)
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:676:51
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:726:13
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:52:16
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:264:21
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:44:16
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:723:17
    at /usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:167:37
    at Timeout._onTimeout (/usr/local/lib/node_modules/mwoffliner/node_modules/async/lib/async.js:652:25)
    at ontimeout (timers.js:429:13)
    at tryOnTimeout (timers.js:289:5)
    at listOnTimeout (timers.js:252:5)
    at Timer.processTimers (timers.js:212:10)
Getting (desktop) article from http://localhost:8000/wiki.installgentoo.com/v3/page/pagebundle/Programming_resources?oldid=34204
[fatal/request][wiki$0/IRC] Not acceptable.
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/1.6.1"
application/json; charset=utf-8; profile="https://www.mediawiki.org/wiki/Specs/pagebundle/2.0.0"
```

**Comment by arlolra** [comment]
> Not acceptable.

This is because you're requesting a version of Parsoid that it is no longer backwards compatible.

https://github.com/openzim/mwoffliner/blob/ae97ec288aee0664f5de365d6f41b3bd2ce10837/lib/Downloader.js#L38

Unfortunately, `1.6.x` should have been `2.x` because it's semantically different, but we messed up on semver.  See https://github.com/wikimedia/parsoid/commit/9c8a72faafc40d407b2d85c6ac5c38baf13af449

See the difference between `1.5.x` (which you're requesting) and `1.6.x` here https://www.mediawiki.org/wiki/Specs/HTML/1.6.x

The most important change being the addition of section elements, which clients newly need to handle.

**Comment by kelson42** [comment]
@arlolra Thx even if I not sure finally which value should be put in `Downloader.js`.
@ISNIT0 Have you all the information you need to fix the pb?

**Comment by ISNIT0** [comment]
I'm not sure yet - I think I'll need to play with it a bit to know.

**Comment by ISNIT0** [comment]
This exception is not a bug in MWOffliner.
The wiki being scraped (`wiki.installgentoo.com`) uses non-standard `wiki` and `api` paths.
The error disappears when the following arguments are added:
`--mwWikiPath="/" --mwApiPath="api.php"`

**Comment by ISNIT0** [comment]
The solution was to use the most recently published version of Parsoid (0.8.1)

**Comment by arlolra** [comment]
> @arlolra Thx even if I not sure finally which value should be put in Downloader.js.

The default for Parsoid `0.9.0` is `1.6.1`, so try with that.  But, as stated above, depending on what you do with the output, you may need special handling for the new section wrapping.

> The solution was to use the most recently published version of Parsoid (0.8.1)

Not sure how that's a solution, since it just updates some dependencies for `0.8.0` because of security advisories.  You'll want to tackle this sooner or later, to continue tracking Parsoid releases.
